### ISSUE TITLE
TTS across servers

### ISSUE THREAD

**Issue by Garry-Marshall** [altoiddealer/ad_discordbot#32]
When the bot is using TTS in voice channel, it should only use TTS when a message was sent in the server where to voice channel resides.
Right now, if you talk to the bot on another server from where it's connected to voice, it'll output to voice across servers.

**Comment by altoiddealer** [comment]
That's very interesting - I'll see about fixing that now.

In the meantime, are you using this bot on multiple servers consecutively?  Just wondering what other insights you may have as I haven't heard of anyone else using it like this (including myself).

**Comment by Garry-Marshall** [comment]
I use it on 3 servers.
The chat history is shared obviously, but it's usually not a big or apparent issue.
It sometimes refers to people by wrong name and such, but that happens on one server as well when multiple people join the conversation. It sometimes continues on a rant from a conversation it had on the other server.
Not a big deal, but when it suddenly starts talking in voice because someone on another server triggered it, that's another story.

If it's technically feasible to separate the chat histories across servers, that would be a big win, but I guess that will not be so straight forward.

**Comment by altoiddealer** [comment]
I just pushed an update that should fix the TTS issue.

Please check and confirm.

In regards to the history management, this will be happening very soon.  Probably within days, actually.

**Comment by altoiddealer** [comment]
Note that I don't currently have a straightforward way to selectively apply / not apply TTS.

To do so would require toggling extensions constantly to apply `disabled` and `enabled`

**Comment by Garry-Marshall** [comment]
You work fast ;-)

I tested it, and the speech is indeed suppressed.
The wav file still gets generated by the extension, and so the reply is slowed down as such but at least it doesn't output it anymore when it mustn't.
Thanks

**Comment by altoiddealer** [comment]
Closing as completed

**Comment by altoiddealer** [comment]
@Garry-Marshall I was thinking about it, and maybe it won't be problematic to toggle the VC

Please test this version of bot.py I published to separate branch, which SHOULD toggle the TTS setting off then back on again.

Please let me know if it consistently works without issues.

https://github.com/altoiddealer/ad_discordbot/blob/a2230445831719871d98cff07907e0eb8773a77b/bot.py

**Comment by Garry-Marshall** [comment]
I ran a quick test, and here's what I found:

The TTS gets turned off correctly when messaging from another server.
But it fails to turn back on from the server where to bot is connected to voice.

**Comment by altoiddealer** [comment]
Thanks - I realized I can just change the condition to test it myself.  I should have this feature available to you today

**Comment by altoiddealer** [comment]
@Garry-Marshall I have another version for you to try out.

https://github.com/altoiddealer/ad_discordbot/blob/toggle_tts/bot.py

If this works expectedly, please also do me a favor and try it out with different "play_modes" (config setting)

If this works as I expect:
- It will still generate TTS for all servers if using `play_mode: 1` or `play_mode: 2` (but will not play on voice channel for mode 2)
- It will toggle off TTS before generating text for non-VC servers.

**Comment by Garry-Marshall** [comment]
Looks like it is now working as intended. Thank you.

As for the play_modes, confirmed following:

play_mode: 0 -> default setting: speaks only on the VC server.
play_mode: 1 -> uploads wav file to any server
play_mode: 2 -> uploads wav file to any server / speaks only on the VC server.

So, works perfect.

**Comment by altoiddealer** [comment]
Beautiful.  I'm going to push this to Main now

**Comment by altoiddealer** [comment]
Pushed the update.

The bug I resolved was to also toggle off TTS generation if `should_send_text` tag is False

**Comment by altoiddealer** [comment]
@Garry-Marshall There's probably a few hidden bugs to work out, but I have the per-channel Histories feature just about wrapped up.

Feel free to try this now, which could be helpful for me to debug anything specifically for multiple servers.

Relavent files are `config.yaml` (new settings), `bot.py` and `modules/database.py` (added a new `/announce` command to set specific channels as Announcement channels)

https://github.com/altoiddealer/ad_discordbot/tree/per_chan_histories

**Comment by Garry-Marshall** [comment]
Hmm, so I tried running your per_chan_histories branch.
I have different nicknames on 2 servers.
I asked it to remember a secret word on server 1 and then asked on server 2 what the word was:
it addressed me by my nickname from server 1 and gave me the secret word.

    per_channel_history: true is set in config file.

**Comment by Garry-Marshall** [comment]
On the second sentence, it explodes:

`19:33:28.250 #1867  ERROR [bot.__main__]: An error occurred in llm_gen(): list index out of range
19:33:28.250 #1869   WARN [bot.__main__]: Note (this may not be the cause of error): "regen" and "continue" commands only work if bot sent message during current session.
Traceback (most recent call last):
  File "E:\Python\oobabooga-windows\bot.py", line 1855, in llm_gen
    last_resp, tts_resp = await loop.run_in_executor(None, process_responses)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\concurrent\futures\thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\bot.py", line 1840, in process_responses
    for resp in chatbot_wrapper(text=llm_payload['text'], state=llm_payload['state'], regenerate=llm_payload['regenerate'], _continue=llm_payload['_continue'], loading_message=True, for_ui=False):
  File "E:\Python\oobabooga-windows\modules\chat.py", line 271, in chatbot_wrapper
    output = apply_extensions('history', output)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\modules\extensions.py", line 231, in apply_extensions
    return EXTENSION_MAP[typ](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\modules\extensions.py", line 125, in _apply_history_modifier_extensions
    history = getattr(extension, "history_modifier")(history)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\extensions\alltalk_tts\script.py", line 280, in history_modifier
    history["visible"][-1][0],
    ~~~~~~~~~~~~~~~~~~^^^^
IndexError: list index out of range`

**Comment by altoiddealer** [comment]
Heya - There's actually been a lot of changes since I shared this with you heh

This will be pushed to main very shortly - and that bug has also been resolves.



You should be quite pleased about this update

**Comment by Garry-Marshall** [comment]
I just tested it against the last commit, and it still bombs out I'm afraid.

**Comment by altoiddealer** [comment]
The error seems to be due to an incompatibility with the `history_modifier` extension.

What does it do?  Can you link to it?

**Comment by altoiddealer** [comment]
Ahh... yeah I see that any extension that modifies history will fail with this.

**Comment by altoiddealer** [comment]
Ahhhhhh! Crap.  Anyway, this may take some more time.

**Comment by Garry-Marshall** [comment]
no problem, no need to rush.
Only extension apart from the defaults, is AllTalk_TTS btw

**Comment by altoiddealer** [comment]
I got that figured out - was much easier bug to squish than I thought it would be.

You can try latest commit which should be 100%

**Comment by Garry-Marshall** [comment]
Nope, still same crash:


`22:22:54.934 #1849  ERROR [bot.__main__]: An error occurred in llm_gen(): list index out of range
22:22:54.934 #1851   WARN [bot.__main__]: Note (this may not be the cause of error): "regen" and "continue" commands only work if bot sent message during current session.
Traceback (most recent call last):
  File "E:\Python\oobabooga-windows\bot.py", line 1837, in llm_gen
    last_resp, tts_resp = await loop.run_in_executor(None, process_responses)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\concurrent\futures\thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\bot.py", line 1822, in process_responses
    for resp in chatbot_wrapper(text=llm_payload['text'], state=llm_payload['state'], regenerate=llm_payload['regenerate'], _continue=llm_payload['_continue'], loading_message=True, for_ui=False):
  File "E:\Python\oobabooga-windows\modules\chat.py", line 271, in chatbot_wrapper
    output = apply_extensions('history', output)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\modules\extensions.py", line 231, in apply_extensions
    return EXTENSION_MAP[typ](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\modules\extensions.py", line 125, in _apply_history_modifier_extensions
    history = getattr(extension, "history_modifier")(history)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\extensions\alltalk_tts\script.py", line 280, in history_modifier
    history["visible"][-1][0],
    ~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range`

**Comment by altoiddealer** [comment]
Try using /reset_conversation in any channels that you used the new bot version.

It is probably trying to use the existing history - the format is bugged.

Or, go into your most recent history log and just add two empty lists into each 'visible' such as 

visible: [
[], []
]

**Comment by Garry-Marshall** [comment]
I tried the /reset_conversation in all channels and reloaded the bot.
Still crashes.

Can you point me to where I can find that history log your referring to?

**Comment by altoiddealer** [comment]
Logs folder in TGWUI folder

**Comment by Garry-Marshall** [comment]
Those files had older timestamps than our tests today.
I cleared them all out to be sure, but it still crashes upon entering the second message to the bot.
(and no log files are being created there)

**Comment by altoiddealer** [comment]
I debugged the issue, and pushed the finalized code to Main.

With per-channel histories, I wanted to keep the logging as light as possible (`internal` and `visible` are basically identical)

The problem was that I did not think `visible` was required - but it was.  

Then, I thought it was resolved because I sent one test message without failure.

Now, I just ensure to add empty strings to `visible` so it is always same length as `internal`.


Long story short, everything should work beautifully now.

Note that there is a new `/announce` command allowing you to set Announce channels for Model Changes and Character changes.

**Comment by Garry-Marshall** [comment]
The crash is indeed fixed now.
As for the missing log files, the setting `autosave_history:` was set to `false`

I ran a couple of tests, and it seems to work as intended, so excellent job there :)

**Comment by altoiddealer** [comment]
Heya, I just added support for per-server Voice Channels, you should be able to run the bot on a voice channel for each server now.

Let me know if it works well!  I only use the bot on one server so don't know 100% that it's working as intended.

**Comment by Garry-Marshall** [comment]
Hey,
I updated the bot and ran a test.
It joins the voice on all servers and generates the wav file for the output, however, it throws a file not found error when trying to upload the wav to the voice channel:

```
Traceback (most recent call last):
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1868, in message_task
    bot_message = await send_responses(params, bot_message)                                 # send responses (text, TTS, images)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1724, in send_responses
    await process_tts_resp(ictx, bot_message)
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 877, in process_tts_resp
    await upload_tts_file(ictx.channel, bot_message)
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 865, in upload_tts_file
    audio = AudioSegment.from_wav(file)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\audio_segment.py", line 808, in from_wav
    return cls.from_file(file, 'wav', parameters=parameters)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\audio_segment.py", line 728, in from_file
    info = mediainfo_json(orig_file, read_ahead_limit=read_ahead_limit)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\utils.py", line 274, in mediainfo_json
    res = Popen(command, stdin=stdin_parameter, stdout=PIPE, stderr=PIPE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] The system cannot find the file specified

23:39:02.482 #1880  ERROR [bot.__main__]: An error occurred while processing "on_message" request: [WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1868, in message_task
    bot_message = await send_responses(params, bot_message)                                 # send responses (text, TTS, images)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1724, in send_responses
    await process_tts_resp(ictx, bot_message)
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 877, in process_tts_resp
    await upload_tts_file(ictx.channel, bot_message)
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 865, in upload_tts_file
    audio = AudioSegment.from_wav(file)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\audio_segment.py", line 808, in from_wav
    return cls.from_file(file, 'wav', parameters=parameters)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\audio_segment.py", line 728, in from_file
    info = mediainfo_json(orig_file, read_ahead_limit=read_ahead_limit)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\pydub\utils.py", line 274, in mediainfo_json
    res = Popen(command, stdin=stdin_parameter, stdout=PIPE, stderr=PIPE)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] The system cannot find the file specified

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\discord\client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1701, in on_message
    await message_task(message, text, 'on_message')
  File "E:\Python\oobabooga-windows\ad_discordbot\bot.py", line 1887, in message_task
    await channel.send(embed=img_gen_embed_info)
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\discord\abc.py", line 1561, in send
    data = await state.http.send_message(channel.id, params=params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\discord\http.py", line 587, in request
    kwargs['data'] = utils._to_json(kwargs.pop('json'))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Python\oobabooga-windows\installer_files\env\Lib\site-packages\discord\utils.py", line 643, in _to_json
    return orjson.dumps(obj).decode('utf-8')
           ^^^^^^^^^^^^^^^^^
TypeError: Type is not JSON serializable: FileNotFoundError

23:39:02.484 #4124   WARN [bot.__main__]: Event error in on_message
```

**Comment by altoiddealer** [comment]
Be sure to update the /modules/ (not just bot.py)

Ideally you’ll have transitioned to the new install method and just use the updater script.

Assuming you were correctly updated, I’ll be looking into this tonight

**Comment by altoiddealer** [comment]
I've invited my bot to another server for testing (I shouldv'e done this awhile ago).

Personally, I'm having no trouble with the bot using TTS on both servers, `play_mode: 2`

Alltalk TTS has an optional parameter to automatically delete audio clips after generation... do you have this enabled?

**Comment by Garry-Marshall** [comment]
Interesting, I went to look into the config, and it hold:
```
play_mode: 2
save_mode: 1
```

When I change play_mode to 0, the error goes away.

**Comment by altoiddealer** [comment]
My config is the same (play mode 2, save mode 1) - no problems.

I recommend checking your local alltalk_tts default settings, automatically delete output could maybe cause the issue...

Otherwise, I think maybe you didn't fully update the bot?

**Comment by Garry-Marshall** [comment]
The git pull from auto update gave warning about the file in modules\text_generation.py being more recent and needing to be stashed first, which is odd as that is not the modules folder from your package, but from the oobabooga parent folder. Anyway I resorted to overwriting all files in ad_discordbot from zip instead, so update-wise, it should be the same.

The config
```
play_mode: 0
save_mode: 1
```
is what I want anyway.
If it works for you, it's something on my end, and for me it works with the config I wanted, so I'm happy.

**Comment by altoiddealer** [comment]
Even if it's working, I recommend getting your installation squared up:

- Back up your current ad_discordbot folder somewhere else.
- in the textgen webui folder, open a CMD and `git clone https://github.com/altoiddealer/ad_discordbot`
- run the updater script one time (will install requirements)
- Now, you can move your backed up user files into the new installation
  - the entire `internal` folder
  - all the `.yaml` files
  - any `sd_outputs`
  - any `user_images`
  - `CMD_FLAGS.txt`

You should now be able to just run the launcher / updater scripts without having to manually download any files, or stashing anything.

**Comment by Garry-Marshall** [comment]
I followed your procedure, and the update is fixed.
However the error remains when I set `play_mode` to `2`.
`play_mode: 0` works fine, and is how I want it set up in any case.

I tried setting `save_mode: 0` to keep the files, but the error remains.
You mentioned an alltaks_tts setting to auto delete.
I went into the config, but it's set as: `"delete_output_wavs": "Disabled"`.

Don't waste any time on this, since it's working for you, it's probably something off in my alltalks setup.

**Comment by altoiddealer** [comment]
I'll certainly be pondering this error 👍 

Thank you for the feedback!
